File: src/2024/Day7.java
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;

public class Day7 {
	public static Object part1(String input) {
		return run(input, false);
	}

	public static Object part2(String input) {
		return run(input, true);
	}

	public static long run(String input, boolean part2) {
		List<Entry<Long, int[]>> equations = input.lines().map(line -> {
			String[] parts = line.split(": ");
			long target = Long.parseLong(parts[0]);
			parts = parts[1].split(" ");
			int[] nums = new int[parts.length];
			for (int i = 0; i < parts.length; i++) {
				nums[i] = Integer.parseInt(parts[i]);
			}

			return Map.entry(target, nums);
		}).toList();

		long sum = 0;
		for (Entry<Long, int[]> equation : equations) {
			if (canSum(equation.getValue(), equation.getKey(), 0, 0, part2)) {
				sum += equation.getKey();
			}
		}

		return sum;
	}

	private static boolean canSum(int[] nums, long target, int index, long currentSum, boolean part2) {
		if(index == nums.length) {
			return currentSum == target;
		}

		if(canSum(nums, target, index + 1, currentSum + nums[index], part2)) {
			return true;
		}

		if(canSum(nums, target, index + 1, currentSum * nums[index], part2)) {
			return true;
		}

		if(part2) {
			int bDigits = (int) (Math.log10(nums[index]) + 1);
			currentSum *= (long) Math.pow(10, bDigits);
			currentSum += nums[index];

			if(canSum(nums, target, index + 1, currentSum, true)) {
				return true;
			}
		}

		return false;
	}

	public static void main(String[] args) throws Throwable {
		String input = Files.readString(Paths.get("inputs/2024/7.txt"));

		long start = System.nanoTime();
		Object result = part1(input);
		long end = System.nanoTime();
		System.out.printf("--- Part 1: %.2fms ---%n", (end - start) / 1e6);
		System.out.println(result);

		start = System.nanoTime();
		result = part2(input);
		end = System.nanoTime();
		System.out.printf("--- Part 2: %.2fms ---%n", (end - start) / 1e6);
		System.out.println(result);
		System.out.println("----------------------");
	}
}

File: src/2024/day7.zig
const std = @import("std");

const ArrayList = std.ArrayList;
const stdout = std.io.getStdOut().writer();
const allocator = std.heap.c_allocator;
const string = []const u8;

const day: u16 = 7;

pub fn part1(input: string) !usize {
    return try run(input, false);
}

pub fn part2(input: string) !usize {
    return try run(input, true);
}

fn run(input: string, p2: bool) !usize {
    const lines = try splitIntoArray(u8, input, "\n");
    var equations = try allocator.alloc(Equation, lines.len);
    for(0..lines.len) |i| {
        const line = lines[i];
        const splitIndex = std.mem.indexOf(u8, line, ": ") orelse return error.@"Invalid input";
        const result = try parseInt(line[0..splitIndex]);

        const numsStr = try splitIntoArray(u8, line[splitIndex + 2..], " ");
        const nums = try allocator.alloc(usize, numsStr.len);
        for(0..numsStr.len) |j| {
            nums[j] = try parseInt(numsStr[j]);
        }

        equations[i] = .{ .target = result, .nums = nums };
    }

    var sum: usize = 0;
    for(equations) |equation| {
        if(try isValid(equation, 1, equation.nums[0], p2)) {
            sum += equation.target;
        }
    }

    return sum;
}

fn isValid(equation: Equation, index: usize, currentSum: usize, p2: bool) !bool {
    if(index == equation.nums.len) {
        return currentSum == equation.target;
    }

    const sum = currentSum + equation.nums[index];
    if(sum <= equation.target and try isValid(equation, index + 1, sum, p2)) {
        return true;
    }

    const product = currentSum * equation.nums[index];
    if(product <= equation.target and try isValid(equation, index + 1, product, p2)) {
        return true;
    }

    if(p2) {
        const combined = concat(currentSum, equation.nums[index]);
        if(combined <= equation.target and try isValid(equation, index + 1, combined, p2)) {
            return true;
        }
    }

    return false;
}

inline fn parseInt(input: string) !usize {
    return try std.fmt.parseInt(usize, input, 10);
}

inline fn concat(first: usize, second: usize) usize {
    var a = first;
    var b = second;
    while(b != 0) {
        a *= 10;
        b /= 10;
    }

    return a + second;
}

const Equation = struct {
    target: usize,
    nums: []usize,
};

fn splitIntoArray(comptime T: type, input: []const T, separator: []const T) ![]const []const T {
    const count = std.mem.count(T, input, separator) + 1;
    const split = try allocator.alloc([]const T, count);

    var itr = std.mem.split(T, input, separator);
    var i: usize = 0;
    while(itr.next()) |token| {
        split[i] = token;
        i += 1;
    }

    return split;
}

pub fn main() !void {
    const path = try std.fmt.allocPrint(allocator, "inputs/2024/{}.txt", .{day});
    const file = try std.fs.cwd().openFile(path, .{ .mode = .read_only });
    defer file.close();
    const input = try file.readToEndAlloc(allocator, (1 << 31) - 1);

    var timer = try std.time.Timer.start();
    const part1Result = try part1(input);
    var time = timer.lap();
    try stdout.print("--- Part 1: {d:.2}ms ---\n", .{(@as(f128, @floatFromInt(time)) / 1_000_000.0)});
    try stdout.print("{any}\n", .{part1Result});

    const part2Result = try part2(input);
    time = timer.lap();
    try stdout.print("--- Part 2: {d:.2}ms ---\n", .{(@as(f128, @floatFromInt(time)) / 1_000_000.0)});
    try stdout.print("{any}\n", .{part2Result});
    try stdout.print("----------------------\n", .{});
}
